docker/k8s/proof-server/proof-server-deployment.yaml

```diff
@@ -0,0 +1,30 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: proof-server-deployment
+spec:
+  replicas: 1  # Initial number of replicas, can be adjusted by HPA
+  selector:
+    matchLabels:
+      app: proof-server
+  template:
+    metadata:
+      labels:
+        app: proof-server
+    spec:
+      containers:
+        - name: proof-server
+          image: argumentcomputer/aptos-light-client:latest
+          imagePullPolicy: IfNotPresent
+          command: [ "sh", "-c", "/app/proof_server --mode \"single\" --addr ${CONTAINER_ADDR}:${CONTAINER_PORT}" ]
+          env:
+            - name: RUST_LOG
+              value: "debug"
+            - name: CONTAINER_ADDR
+              valueFrom:
+                fieldRef:
+                  fieldPath: status.podIP
+            - name: CONTAINER_PORT
+              value: "8080"  # Internal port for the server
+          ports:
+            - containerPort: 8080  # Ensure this matches CONTAINER_PORT
```

Comment:
From what I understood of how HPA works, it seems that the scaling and the routing are both handled by it. We can pass some custom parameters to the configuration to set a "readiness" timer as per the documentation:
> Due to technical constraints, the HorizontalPodAutoscaler controller cannot exactly determine the first time a pod becomes ready when determining whether to set aside certain CPU metrics. Instead, it considers a Pod "not yet ready" if it's unready and transitioned to ready within a short, configurable window of time since it started. This value is configured with the --horizontal-pod-autoscaler-initial-readiness-delay flag, and its default is 30 seconds.

Then the HPA will know when the Pod is ready to be used as a valid replica.

Reply:
Right. I'm not trying to drive the HPA by suggesting to add these readiness and health endpoints.
I'm trying to drive the "available replica count", which will then serve as an input to the HPA (which may then make its own decision).

Readiness and health are K8s-wide concepts:
- Liveness Probes: These are used to indicate if a pod is running. If a liveness probe fails, Kubernetes will kill the pod and attempt to restart it.
- Readiness Probes: These are used to indicate when a pod is ready to start serving traffic. If a readiness probe fails, the pod is removed from the Service's load balancer, meaning it will not receive any traffic until it is ready again.

See https://kubernetes.io/docs/concepts/configuration/liveness-readiness-startup-probes/

Reply:
Got it thank you! I wasn't familiar with those concepts. I'll try a first implementation.